test/server/api_listener_test.cc

```diff
@@ -87,5 +88,46 @@ name: test_api_listener
       "eds_cluster_config {\n    eds_config {\n      path: \"eds path\"\n    }\n  }\n}\n");
 }
 
+TEST_F(ApiListenerTest, HttpApiListenerShutdown) {
+  const std::string yaml = R"EOF(
+name: test_api_listener
+address:
+  socket_address:
+    address: 127.0.0.1
+    port_value: 1234
+api_listener:
+  api_listener:
+    "@type": type.googleapis.com/envoy.config.filter.network.http_connection_manager.v2.HttpConnectionManager
+    stat_prefix: hcm
+    route_config:
+      name: api_router
+      virtual_hosts:
+        - name: api
+          domains:
+            - "*"
+          routes:
+            - match:
+                prefix: "/"
+              route:
+                cluster: dynamic_forward_proxy_cluster
+  )EOF";
+
+  const envoy::config::listener::v3::Listener config = parseListenerFromV2Yaml(yaml);
+
+  auto http_api_listener = HttpApiListener(config, *listener_manager_, config.name());
+
+  ASSERT_EQ("test_api_listener", http_api_listener.name());
+  ASSERT_EQ(ApiListener::Type::HttpApiListener, http_api_listener.type());
+  ASSERT_TRUE(http_api_listener.http().has_value());
+
+  Network::MockConnectionCallbacks network_connection_callbacks;
+  http_api_listener.readCallbacksForTest().connection().addConnectionCallbacks(
+      network_connection_callbacks);
+
+  EXPECT_CALL(network_connection_callbacks, onEvent(Network::ConnectionEvent::LocalClose));
```

Comment:
I spent some time thinking about a better unit test than exposing `readCallbacksForTest` and this expectation but could not come up with anything better. I thought about writing an integration test that waited for reset on the fake stream but that violates #6541.

Reply:
I think the integration test pretty well covers this case so if you don't like this test I would just delete it.

Reply:
I'll leave it with a todo as a reminder that there might be something better out there.